### PR TITLE
Reimplement lazy imports to remove use of codegen

### PR DESCRIPTION
**Pull request by sirosen**
This PR explores some new ideas about how to do our lazy imports.
The initial implementation was revised after a team discussion and review about strategy.

Benchmarking shows nearly equivalent performance to the codegen approach, as the AST traversals are quite fast.

----

The use of '_generate_init.py' makes the code unusual in a few ways which are not key to the lazy import semantics desired in the SDK.
Notably:
- we cannot use our typical code formatters on the init file
- the `__init__.py` cannot be edited by hand -- it must *always* be written via the codegen script
- the codegen script itself contains source-as-strings, which makes those pieces of functionality harder to alter and maintain

A new implementation based around AST-driven introspection provides equivalent semantics without any noticeable performance degradation.

## Initial Implementation

The idea is as follows:
- under the `if t.TYPE_CHECKING` branch, write relative imports as usual
- outside of that branch, provide a canonical `__all__` list to drive behaviors and identify "known names"
- in `__getattr__`, library-ize most of the lazy loading logic
- in that lazy-loading library, inspect module source and read the `if t.TYPE_CHECKING` branch
  - cache aggressively to avoid redoing AST work
  - scan it for `from .x import y` statements
  - redrive those parsed imports through `importlib`

## Revised Implementation

Following the above implementation, we took the same idea and tried using `.pyi` files to refine it further.
The ideas in the refinement are:
- we can load data out of an `__init__.pyi` and thereby avoid having to identify the `if t.TYPE_CHECKING` branch
- the internal lazy loading library can provide the implementations of `__dir__` and `__getattr__`, as well as the `__all__` tuple (for runtime use)
- now that we're locking in the implementation, more vigorous documentation effort is now appropriate

----
📚 Documentation preview 📚: https://globus-sdk-python--1100.org.readthedocs.build/en/1100/

### CODE REVIEW

**Comment by sirosen** [comment]
I've done a bit more testing and found that the original timing test I did was not correct.
I was timing how long it took to access an attribute of `globus_sdk`, but not clearing the module attribute for each run (so I was not _actually_ timing the codepath). Once corrected, the codegen approach properly shows itself to be faster, but note that this is down at nanoseconds scale.

---

Here's the simple benchmark script:
```python
import timeit

N = 10000

measurements = timeit.repeat(
    """
if hasattr(globus_sdk, "TransferClient"):
    del globus_sdk.TransferClient

globus_sdk.TransferClient""",
    setup="""
import globus_sdk
""",
    number=N,
)
print(f"benchmark: {min(measurements) / N * 1000_000_000}ns")
```
`benchmark.py`

Selecting the min of three runs for each implementation, I get:
- current (codegen): `benchmark: 1944.7158090770245ns`
- PR (AST): `benchmark: 11011.32296025753ns`

Both are exactly the same after the attribute is loaded (if you don't delete it).

In other words, the new approach is about 9000ns slower, or 0.009ms slower. I think that's an acceptable performance regression in order for us to have a more normal, and therefore maintainable, `__init__.py`.

**Comment by sirosen** [comment]
I've pushed the new implementation. With the new behavior, the benchmark above runs similarly:
```
benchmark: 10612.473799847066ns
```

**Review by kurtmckee — APPROVED** [review]

File: src/globus_sdk/_lazy_import.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+i.e. Given
+
+  foo.py
+  foo.pyi
+  bar.py
+
+then if `foo.pyi` has an import `from .bar import BarType`, it is possible to
+*read* `foo.pyi` at runtime and use that information to load
+`BarType` from `bar`.
```
[the remark this answers is not in the document]
Comment (sirosen) [reply src/globus_sdk/_lazy_import.py:16]: I'd rather tackle this now if we can. I added a rationale doc comment. I'll wait for feedback before I consider merging. :+1:

**Review by kurtmckee — APPROVED** [review]
